pkg/storage/iterators.go

```diff
@@ -199,3 +199,67 @@ func NewFilteredTupleKeyIterator(iter TupleKeyIterator, filter TupleKeyFilterFun
 		filter,
 	}
 }
+
+// TupleKeyConditionFilterFunc is a filter function that is used to filter out
+// tuples from a [TupleKeyIterator] that don't meet the tuple the conditions provided by the request.
+// Implementations should return true if the tuple should be returned
+// and false if it should be filtered out.
+// Errors will be treated as false. The iterator can return the last seen error before ErrIteratorDone.
+type TupleKeyConditionFilterFunc func(tupleKey *openfgav1.TupleKey) (bool, error)
+
+type conditionsFilteredTupleKeyIterator struct {
+	iter      TupleKeyIterator
+	filter    TupleKeyConditionFilterFunc
+	lastError error
+	isClosed  bool
+	onceValid bool
+}
+
+var _ TupleKeyIterator = &conditionsFilteredTupleKeyIterator{}
+
+// Next returns the next most tuple in the underlying iterator that meets
+// the filter function this iterator was constructed with.
+func (f *conditionsFilteredTupleKeyIterator) Next(ctx context.Context) (*openfgav1.TupleKey, error) {
+	if f.isClosed {
+		return nil, ErrIteratorDone
+	}
+	for {
+		tuple, err := f.iter.Next(ctx)
+		if err != nil {
+			if !errors.Is(err, ErrIteratorDone) {
+				f.isClosed = true
+				if f.onceValid {
+					return nil, err
+				}
+				return nil, f.lastError
+			}
+			return nil, err
+		}
+
+		valid, err := f.filter(tuple)
+		if err != nil {
+			f.lastError = err
+			continue
+		}
+		if !valid {
+			continue
+		}
+		f.onceValid = true
+		return tuple, nil
+	}
+}
+
+// Stop see [Iterator.Stop].
+func (f *conditionsFilteredTupleKeyIterator) Stop() {
+	f.iter.Stop()
+	f.isClosed = true
+}
+
+// NewConditionsFilteredTupleKeyIterator returns a [TupleKeyIterator] that filters out all
+// [*openfgav1.Tuple](s) that don't meet the conditions of the provided [TupleKeyFilterFunc].
+func NewConditionsFilteredTupleKeyIterator(iter TupleKeyIterator, filter TupleKeyConditionFilterFunc) TupleKeyIterator {
```

Comment:
Instead of returning an interface here why don't we return the explicit type *storage.ConditionsFilteredTupleKeyIterator? https://100go.co/#returning-interfaces-7

I'd also change the signature of `checkUsersetFastPath` to accept a `ConditionsFilteredTupleKeyIterator`. In the future I don't want to make the mistake of accidentally forgetting to perform all these validations prior to working with a tuple.

Reply:
I would defer this to a separate PR as all iterators in storage/iterators.go are private type (and only the interface is exposed).  If we are moving to this route, we should do it for all iterators, not just ConditionsFilteredTupleKeyIterator.

Reply:
This PR is the right place to make that change, IMO. We are introducing a new object, we should ensure people don't misuse it right in this PR

Reply:
Done